k4FWCore/components/IOSvc.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2014-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "IOSvc.h"
+
+#include "podio/Frame.h"
+#include "podio/FrameCategories.h"
+
+#include "k4FWCore/FunctionalUtils.h"
+#include "k4FWCore/KeepDropSwitch.h"
+
+#include "GaudiKernel/AnyDataWrapper.h"
+#include "GaudiKernel/IEventProcessor.h"
+
+#include <mutex>
+#include <tuple>
+
+StatusCode IOSvc::initialize() {
+  StatusCode sc = Service::initialize();
+  if ( !sc.isSuccess() ) {
+    error() << "Unable to initialize base class Service." << endmsg;
+    return sc;
+  }
+  if (!m_readingFileNames.empty()) {
+    m_reader = std::make_unique<podio::ROOTReader>();
+    try {
+      m_reader->openFiles(m_readingFileNames);
+    } catch (std::runtime_error& e) {
+      error() << "Error when opening files: " << e.what() << endmsg;
+      throw e;
+    }
+    m_entries = m_reader->getEntries(podio::Category::Event);
+  }
+
+  m_switch = KeepDropSwitch(m_outputCommands);
+
+  m_incidentSvc = service("IncidentSvc");
+  if (!m_incidentSvc) {
+    error() << "Unable to locate IIncidentSvc interface" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  m_incidentSvc->addListener(this, IncidentType::EndEvent);
+
+  m_dataSvc = service("EventDataSvc");
+  if (!m_dataSvc) {
+    error() << "Unable to locate IDataSvc interface" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  m_hiveWhiteBoard = service("EventDataSvc");
+
+  return Service::initialize();
```

Comment:
`Service::initialize` is called twice. Should be called only at the beginning
```suggestion
  StatusCode sc = Service::initialize();
  if ( !sc.isSuccess() ) {
    error() << "Unable to initialize base class Service." << endmsg;
    return sc;
  }
  if (!m_readingFileNames.empty()) {
    m_reader = std::make_unique<podio::ROOTReader>();
    try {
      m_reader->openFiles(m_readingFileNames);
    } catch (std::runtime_error& e) {
      error() << "Error when opening files: " << e.what() << endmsg;
      throw e;
    }
    m_entries = m_reader->getEntries(podio::Category::Event);
  }

  m_switch = KeepDropSwitch(m_outputCommands);

  m_incidentSvc = service("IncidentSvc");
  if (!m_incidentSvc) {
    error() << "Unable to locate IIncidentSvc interface" << endmsg;
    return StatusCode::FAILURE;
  }
  m_incidentSvc->addListener(this, IncidentType::EndEvent);

  m_dataSvc = service("EventDataSvc");
  if (!m_dataSvc) {
    error() << "Unable to locate IDataSvc interface" << endmsg;
    return StatusCode::FAILURE;
  }

  m_hiveWhiteBoard = service("EventDataSvc");

  return StatusCode::SUCCESS;
```